lib/types.nix

```diff
@@ -105,6 +105,17 @@ rec {
     binOp   = a: b: null;
   };
 
+
+  isSecret = isType "secret";
+  mkSecret = path:
+    assert lib.assertMsg (types.path.check path)
+      "lib.mkSecret: argument must be a path";
+    {
+      _type = "secret";
+      _secret = path;
+    };
```

Comment:
> This looks promising.
> 
> I suppose this is also draft because it's missing
> 
>  * [ ]  tests, all of
>  ...
>  * [ ]  documentation
>  ...

Yes, of course.

> If you don't mind me dumping my brain here, these are some topics that need answers in order to _validate_ the solution. Those answers need not be implementations; just outlines of designs.
> 
> * Support secrets in the evaluator, such as with `deployment.keys` in NixOps. This currently relies on `lib.generators` to work, as `lib.generators` does not write to the store.
> 
> * Push secrets, as opposed to pulling from files
> 
> * Reading from commands, environment variables, systemd `LoadCredentials`
> 
> * ramfs
> 
> * support various sources, such as `nixos-rebuild` (secrets stored in e.g. `/root`?), `agenix`, `sops-nix`, etc - pluggable?

I'm not sure what the first items in the list mean, but sourcing the password from a custom command rather than just a file sounds interesting and certainly doable.

> @infinisil has done work similar work in
>
>   - https://github.com/NixOS/nixpkgs/pull/78640
>
> The name `secretPath` *may* be better, depending on how we want this type to evolve. If we add support for secrets from a command invocation, `secret` is better.

Indeed - I'll have to look more into the prior PRs. I like the idea of making the `secret` type more flexible, rather than renaming it `secretPath`.